sdks/python/apache_beam/io/filebasedsource.py

```diff
@@ -377,8 +382,12 @@ def process(self, element, *args, **kwargs):
     if not source_list:
       return
     source = source_list[0].source
-    for record in source.read(range.new_tracker()):
-      yield record
+
+    if self._with_filename:
+      return [(metadata.path, record)
+              for record in source.read(range.new_tracker())]
+    else:
+      return [record for record in source.read(range.new_tracker())]
```

Comment:
Why are we changing yield to returning a list for the existing use case (without filename). Does this provide any advantages?

Reply:
Yeah, I think this could still result in OOMs (we are putting all elements to a list before returning).